tokenserver-auth/src/crypto.rs

```diff
@@ -0,0 +1,135 @@
+use hkdf::Hkdf;
+use hmac::{Hmac, Mac};
+use jsonwebtoken::{errors::ErrorKind, jwk::Jwk, Algorithm, DecodingKey, Validation};
+use serde::de::DeserializeOwned;
+use sha2::Sha256;
+use tokenserver_common::TokenserverError;
+pub const SHA256_OUTPUT_LEN: usize = 32;
+/// A triat representing all the required cryptographic operations by the token server
+pub trait Crypto {
+    type Error;
+    /// HKDF key derivation
+    fn hkdf(&self, secret: &str, salt: Option<&[u8]>, info: &[u8]) -> Result<Vec<u8>, Self::Error>;
+
+    /// HMAC signiture
+    fn hmac_sign(&self, key: &[u8], payload: &[u8]) -> Result<Vec<u8>, Self::Error>;
+
+    /// Verify an HMAC signature on a payload given a shared key
+    fn hmac_verify(&self, key: &[u8], payload: &[u8], signature: &[u8]) -> Result<(), Self::Error>;
+}
+
+/// An implementation for the needed cryptographic using
+///    the hmac crate for hmac and hkdf crate for hkdf
+pub struct CryptoImpl {}
+
+impl Crypto for CryptoImpl {
+    type Error = TokenserverError;
+    fn hkdf(&self, secret: &str, salt: Option<&[u8]>, info: &[u8]) -> Result<Vec<u8>, Self::Error> {
+        let hk = Hkdf::<Sha256>::new(salt, secret.as_bytes());
+        let mut okm = [0u8; SHA256_OUTPUT_LEN];
+        hk.expand(info, &mut okm)
+            .map_err(|_| TokenserverError::internal_error())?;
+        Ok(okm.to_vec())
+    }
+
+    fn hmac_sign(&self, key: &[u8], payload: &[u8]) -> Result<Vec<u8>, Self::Error> {
+        let mut mac: Hmac<Sha256> =
+            Hmac::new_from_slice(key).map_err(|_| TokenserverError::internal_error())?;
+        mac.update(payload);
+        Ok(mac.finalize().into_bytes().to_vec())
+    }
+
+    fn hmac_verify(&self, key: &[u8], payload: &[u8], signature: &[u8]) -> Result<(), Self::Error> {
+        let mut mac: Hmac<Sha256> =
+            Hmac::new_from_slice(key).map_err(|_| TokenserverError::internal_error())?;
+        mac.update(payload);
+        mac.verify_slice(signature)
+            .map_err(|_| TokenserverError::internal_error())?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum OAuthVerifyError {
```

Comment:
Should we add a `ReportableError` trait to this so that we can get metrics? 

Looks like this gets eaten by `VerifyToken::verify()`, but might be good to track a few errors to catch any anomalies. (e.g. `InvalidKey` or `DecodingError`)

Reply:
That's a great idea! One thing I wanted to do is track why tokenserver is seemingly hitting FxA often, and tracking those errors will get us exactly there - I'll give adding that trait a shot